gms/gossiper.cc

```diff
@@ -421,6 +425,120 @@ future<> gossiper::handle_echo_msg() {
     return make_ready_future<>();
 }
 
+// Must run inside a seastar thread
+uint64_t gossiper::get_gossip_status_token_hash() const {
+    auto nodes = boost::copy_range<std::set<gms::inet_address>>(endpoint_state_map | boost::adaptors::map_keys);
+    if (nodes.empty()) {
+        return 0;
+    }
+    xx_hasher h;
+    bytes_ostream out;
+    for (const auto& node : nodes) {
+        thread::maybe_yield();
+        out.clear();
+        auto state = get_application_state_ptr(node, application_state::TOKENS);
+        auto tokens = state ? state->value : "";
+        ser::serialize(out, tokens);
+        ser::serialize(out, get_gossip_status(node));
+        ser::serialize(out, node);
+        feed_hash(h, out.linearize());
+    }
+    return h.finalize_uint64();
+}
+
+static future<cluster_ring_status_response>
+get_cluster_ring_status(gossiper& g) {
+    using response = cluster_ring_status_response;
+    return g.container().map([] (gossiper& g) -> future<response> {
+        return seastar::async([&g] {
+            auto& tm = service::get_local_storage_service().get_token_metadata();
+            cluster_ring_status_response resp = {{
+                {response::nr_known_nodes, format("{}", g.endpoint_state_map.size())},
+                {response::status_token_hash, format("{}", g.get_gossip_status_token_hash())},
+                {response::normal_token_hash, format("{}", tm.get_normal_token_hash())},
+                {response::bootstrap_token_hash, format("{}", tm.get_bootstrap_token_hash())},
+                {response::leaving_nodes_hash, format("{}", tm.get_leaving_endpoints_hash())},
+                {response::pending_ranges_hash, format("{}", tm.get_pending_ranges_hash())}
+            }};
+            return resp;
+        });
+    }).then([] (std::vector<cluster_ring_status_response> results) {
+        bool same_results = std::adjacent_find(results.begin(), results.end(),
+                std::not_equal_to<cluster_ring_status_response>()) == results.end();
+        results.front().status[response::same_on_all_shards] = same_results ?
+                response::same_on_all_shards_yes : response::same_on_all_shards_no;
+        return results.front();
+    });
+}
+
+// Send CLUSTER_RING_STATUS RPC verb to remote nodes to query the cluster
+// status. If local node has the same cluster ring status as the remote nodes,
+// it means the cluster has the same view on the cluster ring status.
+static future<> wait_for_cluster_ring_status_to_match(gossiper& g) {
+    return seastar::async([&g] {
+        while (true) {
```

Comment:
Shouldn't we have a timeout to avoid infinite loop?

Reply:
The rationale is that all acceptable exits case should be handled explicitly. If we do not exit, then it has to match. If not, there is something wrong with the cluster, it is better not start. Fix the issue, then allow the start.

Reply:
Ok.